### PR TITLE
Update `qiskit.utils.wrap_method` for Python 3.11

### LINKED ISSUE #9291: `qiskit.test.decorators.enforce_subclasses_call` machinery is broken in 3.11.1
### Environment

- **Qiskit Terra version**: 

HEAD of main, currently `dd7f9390cf07`

- **Python version**: 

3.11.1

- **Operating system**: 

Both macOS and ubuntu, so far.

### What is happening?

`import test.python` raises an error in 3.11.1, therefore tests do not run. @wshanks has indicated that it is working fine for him in 3.11.0 but not 3.11.1 (I don't have 3.11.0 installed).

```
Python 3.11.1 (v3.11.1:a7a450f84a, Dec  6 2022, 15:24:06) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import test.python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ihincks/ibm/qiskit-terra/test/__init__.py", line 16, in <module>
    from qiskit.test.utils import generate_cases
  File "/Users/ihincks/ibm/qiskit-terra/qiskit/test/__init__.py", line 15, in <module>
    from .base import QiskitTestCase
  File "/Users/ihincks/ibm/qiskit-terra/qiskit/test/base.py", line 167, in <module>
    class QiskitTestCase(BaseQiskitTestCase):
  File "/Users/ihincks/ibm/qiskit-terra/qiskit/utils/classtools.py", line 117, in out
    retval = method(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^
TypeError: TestCase.__init_subclass__() missing 1 required positional argument: 'cls'
```

Or, see an example of this happening in CI: https://github.com/Qiskit/qiskit-experiments/actions/runs/3704009805/jobs/6281449337

### How can we reproduce the issue?

Run the tests or simply `import test.python` in Python 3.11.1.

### What should happen?

This should not error.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by wshanks**
### Summary

Fixed the behavior of `qiskit.utils.wrap_method` when applied to a `classmethod` that is on `type`. Code changes in Python 3.11 and 3.11.1 revealed that `wrap_method` did not avoid calling a `classmethod` desscriptor's `__get__` method. `wrap_method` was changed to use `inspect.getattr_static` to avoid calling `__get__`.

### Details and comments

The poor handling of `classmethod` by `wrap_method` was first noticed with Python 3.11.1 as described in #9291. `qiskit.test.decorators.enforce_subclasses_call` wraps `QiskitBaseTestCase.__init_subclass__` using `wrap_method`. Prior to 3.11.1, this meant wrapping `object.__init_subclass__` (which is a builtin method that behaves differently), but in 3.11.1 `unittest.TestCase` added an `__init_subclass__` method. This change meant that now `wrap_method` was wrapping a "real" `classmethod` (`__init_subclass__` is a special case but it behaves enough like a `classmethod` for this issue).

In Python 3.11.0, the behavior of the bound method descriptor changed (specifically, it was [this PR](https://github.com/python/cpython/pull/31367) and the change to `funcobject.c` there). The following code illustrates the difference in behavior:

```
class A:
    def m(self):
        pass


print(A().m.__get__(None, A))
```

In Python 3.10, this gives

    <bound method A.m of <__main__.A object at 0x7f942bc8ace0>>

while in Python 3.11 it gives

    <function A.m at 0x7f432b0daa20>

So in Python 3.10 and earlier, the bound method's descriptor returns itself. With Python 3.11 though, the descriptor returns the unbound function.

Prior to the change here, `wrap_method` used `object.__getattribute__` and `type.__getattribute__` to retrieve the descriptor for the method being wrapped so that its `__get__` could be  called by the wrapping descriptor. For methods that are not in `dir(type)`, it used `object.__getattribute__(cls, name)` to retrieve the descriptor in a loop over every `cls` in the wrapped class's MRO passing on `AttributeError`. `object.__getattribute__(cls, name)` checks `type(cls)` (so `type`) for a descriptor and otherwise checks `cls.__dict__` for `name`, raising an `AttributeError` if `name` is not found (the code is [here](https://github.com/python/cpython/blob/432117cd1f59c76d97da2eaff55a7d758301dbc7/Objects/object.c#L1231)). By doing this loop over the MRO, `name` is eventually found. Note, however, that `object.__getattribute__` follows a different path if `name` is a descriptor in `type(cls)` (so in `type`). To avoid this different path, `wrap_method` checked `name` against `dir(type)` and used `type.__getattribute__` instead. `type.__getattribute__(cls, name)` loops over the MRO of `cls` looking for `name` and when it finds it  calls `.__get__(None, cls)` on it (the code is [here](https://github.com/python/cpython/blob/73d2b15bd278e84e89de83dbd2ee64eac2e63cda/Objects/typeobject.c#L3902)). This call returns a bound method version of the classmethod with `cls` bound and this is where the Python 3.11 change in behavior comes in. In `qiskit.utils.classtools._WrappedMethod`, `__get__` tries to delegate to the wrapped method by calling `.__get__` on it  but for this `type` case that means calling `.__get__` on a bound method instead of the `classmethod` and so for 3.11 getting back an unbound function that expects an extra first argument containing the `cls`.

So in Python 3.11.1 with the addition of `__init_subclass__` to `unittest.TestCase`, the `wrap_method` call around `__init_subclass__` on `QiskitBaseTestCase` meant that the definition of `QiskitTestCase(QiskitBaseTestCase)` now tried to  call the wrapped `__init_subclass__` as an unbound function without passing it the required `cls` argument and this led to the error. In 3.11.0, the handling of the descriptor did not matter because `object.__init_subclass__` was the method being wrapped and it has a `*args, **kwargs` signature and does nothing, so it did not matter that it was invoked incorrectly (see [here](https://github.com/python/cpython/blob/48e352a2410b6e962d40359939a0d43aaba5ece9/Objects/typeobject.c#L5863-L5866)).

Closes #9291

### CODE REVIEW

**Comment by Cryoris** [comment]
Nice this seems to work! The 3.11 tests are failing due to NumPy, but that should be fixed now.

File: releasenotes/notes/wrap-method-311-147d254d4b40e805.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+fixes:
+  - |
+    Fixed bad handling of ``__init_subclass__`` by `~qiskit.utils.wrap_method`
```
Comment (Cryoris) [inline releasenotes/notes/wrap-method-311-147d254d4b40e805.yaml:4]: If we want the releasenotes to link this correctly I think we need to mark this as a function
```suggestion
    Fixed bad handling of ``__init_subclass__`` by :func:`~qiskit.utils.wrap_method`
```

File: releasenotes/notes/wrap-method-311-147d254d4b40e805.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    in Python 3.11. Previously, in Python 3.11, ``wrap_method`` would wrap the
+    unbounded function of ``__init_subclass__`` and then fail when invoked
+    because the bound class object was not passed to the function. Starting in
+    Python 3.11.1, this issue affected `~qiskit.test.QiskitTestCase`,
```
Comment (wshanks) [inline releasenotes/notes/wrap-method-311-147d254d4b40e805.yaml:8]:
```suggestion
    Python 3.11.1, this issue affected :class:`~qiskit.test.QiskitTestCase`,
```

File: qiskit/utils/classtools.py [hunk 3]
```diff
@@ -28,7 +28,7 @@
 # `type` itself has several methods (mostly dunders).  When we are wrapping those names, we need to
 # make sure that we don't interfere with `type.__getattribute__`'s handling that circumvents the
 # normal inheritance rules when appropriate.
-_TYPE_METHODS = set(dir(type))
+_TYPE_METHODS = set(dir(type)) - {"__init_subclass__"}
```
Comment (wshanks) [inline qiskit/utils/classtools.py:31]: I have been thinking about if there is a comment that should go with the change here, but I am not sure if I know how `__init_subclass__` is handled well enough to write that comment. I was also trying to decide if `__init_subclass__` was the only thing that should go here. It's at least the only thing that matters to the tests.

**Comment by wshanks** [comment]
While this lets the tests run, I think it is not quite right in general.

**Comment by wshanks** [comment]
Now I think the change is correct.

**Review by mtreinish — APPROVED** [review]
This LGTM, I didn't know about `getattr_static` but that looks like exactly what we were trying to hack together before (and what was causing issues for us on 3.11.1). I'm going to hold off on automerge until @jakelishman can take a look though since he wrote all of this decorator code (and I've lost most of the context I previously had for it).

**Review by jakelishman — APPROVED** [review]
This looks great, thanks Will.  I think I didn't know about `getattr_static` when I wrote that first implementation, which is part of why it got super hacky.

**Pull request merged**